test/e2e/framework/pod/wait.go

```diff
@@ -65,31 +66,68 @@ const (
-func errorBadPodsStates(badPods []v1.Pod, desiredPods int, ns, desiredState string, timeout time.Duration, err error) string {
-	errStr := fmt.Sprintf("%d / %d pods in namespace %q are NOT in %s state in %v\n", len(badPods), desiredPods, ns, desiredState, timeout)
-	if err != nil {
-		errStr += fmt.Sprintf("Last error: %s\n", err)
+type timeoutError struct {
+	msg string
+}
+
+func (e *timeoutError) Error() string {
+	return e.msg
+}
+
+func TimeoutError(format string, args ...interface{}) *timeoutError {
+	return &timeoutError{
+		msg: fmt.Sprintf(format, args...),
+	}
+}
+
+// MaybeTimeoutError returns a TimeoutError if err is a timeout. Otherwise, wrap err.
+// taskFormat and taskArgs should be the task being performed when the error occurred,
+// e.g. "waiting for pod to be running".
+func MaybeTimeoutError(err error, taskFormat string, taskArgs ...interface{}) error {
+	if IsTimeout(err) {
+		return TimeoutError("timout "+taskFormat, taskArgs...)
+	} else if err != nil {
+		return fmt.Errorf("error while %s: %w", fmt.Sprintf(taskFormat, taskArgs...), err)
+	} else {
+		return nil
 	}
+}
+
+func IsTimeout(err error) bool {
+	if errors.As(err, &timeoutError{}) {
+		return true
+	}
+	return errors.Is(err, wait.ErrWaitTimeout)
+}
+
+// errorBadPodsStates create error message of basic info of bad pods for debugging.
+func errorBadPodsStates(badPods []v1.Pod, desiredPods int, ns, desiredState string, timeout time.Duration, err error) error {
+	errStr := fmt.Sprintf("%d / %d pods in namespace %s are NOT in %s state in %v\n", len(badPods), desiredPods, ns, desiredState, timeout)
+
 	// Print bad pods info only if there are fewer than 10 bad pods
 	if len(badPods) > 10 {
-		return errStr + "There are too many bad pods. Please check log for details."
+		errStr += "There are too many bad pods. Please check log for details."
+	} else {
+		buf := bytes.NewBuffer(nil)
+		w := tabwriter.NewWriter(buf, 0, 0, 1, ' ', 0)
+		fmt.Fprintln(w, "POD\tNODE\tPHASE\tGRACE\tCONDITIONS")
+		for _, badPod := range badPods {
+			grace := ""
+			if badPod.DeletionGracePeriodSeconds != nil {
+				grace = fmt.Sprintf("%ds", *badPod.DeletionGracePeriodSeconds)
+			}
+			podInfo := fmt.Sprintf("%s\t%s\t%s\t%s\t%+v",
+				badPod.ObjectMeta.Name, badPod.Spec.NodeName, badPod.Status.Phase, grace, badPod.Status.Conditions)
+			fmt.Fprintln(w, podInfo)
+		}
+		w.Flush()
+		errStr += buf.String()
 	}
 
-	buf := bytes.NewBuffer(nil)
-	w := tabwriter.NewWriter(buf, 0, 0, 1, ' ', 0)
-	fmt.Fprintln(w, "POD\tNODE\tPHASE\tGRACE\tCONDITIONS")
-	for _, badPod := range badPods {
-		grace := ""
-		if badPod.DeletionGracePeriodSeconds != nil {
-			grace = fmt.Sprintf("%ds", *badPod.DeletionGracePeriodSeconds)
-		}
-		podInfo := fmt.Sprintf("%s\t%s\t%s\t%s\t%+v",
-			badPod.ObjectMeta.Name, badPod.Spec.NodeName, badPod.Status.Phase, grace, badPod.Status.Conditions)
-		fmt.Fprintln(w, podInfo)
+	if err != nil && !IsTimeout(err) {
+		return fmt.Errorf("%s\nLast error: %w", errStr, err)
```

Comment:
Note to reviewers: The changes to this function are to ensure that the last error is wrapped correctly, so that it is identifiable with errors.As/Is